### ISSUE TITLE
cluster_slots() Error

### ISSUE THREAD

**Issue by c-buisson** [Grokzen/redis-py-cluster#163]
Hello,

I am seeing this error whenever I try to use `cluster_slots()`.
I have a Redis cluster (3.2) with 3 masters and 3 slaves.

Have you experienced this before?
Thanks

 From a Python3 shell:

```
import rediscluster
client = rediscluster.StrictRedisCluster(startup_nodes=[{"host": "localhost", "port":10000}])
client.cluster_slots()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/cbuisson/env/local/lib/python3.4/site-packages/rediscluster/client.py", line 545, in cluster_slots
    return self.execute_command('CLUSTER SLOTS')
  File "/home/cbuisson/env/local/lib/python3.4/site-packages/rediscluster/utils.py", line 101, in inner
    return func(*args, **kwargs)
  File "/home/cbuisson/env/local/lib/python3.4/site-packages/rediscluster/client.py", line 269, in execute_command
    return self._execute_command_on_nodes(node, *args, **kwargs)
  File "/home/cbuisson/env/local/lib/python3.4/site-packages/rediscluster/client.py", line 353, in _execute_command_on_nodes
    res[node["name"]] = self.parse_response(connection, command, **kwargs)
  File "/home/cbuisson/env/local/lib/python3.4/site-packages/redis/client.py", line 587, in parse_response
    return self.response_callbacks[command_name](response, **options)
  File "/home/cbuisson/env/local/lib/python3.4/site-packages/rediscluster/utils.py", line 136, in parse_cluster_slots
    'master': fix_server(*master),
TypeError: fix_server() takes 2 positional arguments but 3 were given
```

**Comment by Grokzen** [comment]
@c-buisson What version of the client are you running? Is this something that hajve appeared in the latest relese `1.3.1`??

**Comment by c-buisson** [comment]
I noticed this with `1.3.0` and `1.3.1`.
Were you able to reproduced this error?

**Comment by Grokzen** [comment]
@c-buisson This is not normal behavior from the client or the cluster. There seems to be a problem with the cluster you are running on your end.

What version of redis (exactly) are you running in your cluster? Can you connect to your cluster on the host you try to connect to in startup_nodes (localhost:10000) and run `CLUSTER SLOTS` and paste the output back to me, it should help me to track down why the parsing i behaving badly. Also what version of redis-py?? Where and how are you running your cluster? Are you using something like docker or running it in some other kind of virtualization or directly on your machine?

If you have the error reproducible please paste what is in the `master` variable just before the last line in the exception. I can't really reproduce the problem unless i know these details.

**Comment by Grokzen** [comment]
My guess is that your cluster is probably setup wrong and that cause these problems.

**Comment by c-buisson** [comment]
I am using redis-server `3:3.2.4-1~bpo8+1` and redis-py version is: `2.10.5`

Here is the output of `CLUSTER SLOTS` from redis-cli:
```
cbuisson@rediscluster01:~ ()$ redis-cli -p 10000
127.0.0.1:10000> cluster slots
1) 1) (integer) 5461
   2) (integer) 10922
   3) 1) "10.0.0.1"
      2) (integer) 10000
      3) "3588b4cf9fc72d57bb262a024747797ead0cf7ea"
   4) 1) "10.0.0.4"
      2) (integer) 10000
      3) "a72c02c7d85f4ec3145ab2c411eefc0812aa96b0"
2) 1) (integer) 10923
   2) (integer) 16383
   3) 1) "10.0.0.2"
      2) (integer) 10000
      3) "ffd36d8d7cb10d813f81f9662a835f6beea72677"
   4) 1) "10.0.0.5"
      2) (integer) 10000
      3) "5c15b69186017ddc25ebfac81e74694fc0c1a160"
3) 1) (integer) 0
   2) (integer) 5460
   3) 1) "10.0.0.3"
      2) (integer) 10000
      3) "069cda388c7c41c62abe892d9e0a2d55fbf5ffd5"
   4) 1) "10.0.0.6"
      2) (integer) 10000
      3) "dc152a08b4cf1f2a0baf775fb86ad0938cb907dc"
```

`CLUSTER INFO`
```
127.0.0.1:10000> cluster info
cluster_state:ok
cluster_slots_assigned:16384
cluster_slots_ok:16384
cluster_slots_pfail:0
cluster_slots_fail:0
cluster_known_nodes:6
cluster_size:3
cluster_current_epoch:11
cluster_my_epoch:1
cluster_stats_messages_sent:4667794
cluster_stats_messages_received:4667505
```

`CLUSTER NODES`
```
127.0.0.1:10000> cluster nodes
3588b4cf9fc72d57bb262a024747797ead0cf7ea 10.0.0.1:10000 master - 0 1476902039357 2 connected 5461-10922
5c15b69186017ddc25ebfac81e74694fc0c1a160 10.0.0.5:10000 slave ffd36d8d7cb10d813f81f9662a835f6beea72677 0 1476902038990 11 connected
a72c02c7d85f4ec3145ab2c411eefc0812aa96b0 10.0.0.4:10000 slave 3588b4cf9fc72d57bb262a024747797ead0cf7ea 0 1476902039724 5 connected
dc152a08b4cf1f2a0baf775fb86ad0938cb907dc 10.0.0.6:10000 slave 069cda388c7c41c62abe892d9e0a2d55fbf5ffd5 0 1476902040822 8 connected
ffd36d8d7cb10d813f81f9662a835f6beea72677 10.0.0.2:10000 master - 0 1476902039906 11 connected 10923-16383
069cda388c7c41c62abe892d9e0a2d55fbf5ffd5 10.0.0.3:10000 myself,master - 0 0 1 connected 0-5460
```

The cluster is running on really dedicated hardware. No virtualization nor Docker here.

I am able to reproduce this error. I have added a print statement to print out the `master` variable in `utils.py`:
```
>>> r.cluster_slots()
MASTER: ['10.0.0.3', 10000L, '069cda388c7c41c62abe892d9e0a2d55fbf5ffd5']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/cbuisson/env-cache-benchmark/local/lib/python2.7/site-packages/rediscluster/client.py", line 549, in cluster_slots
    return self.execute_command('CLUSTER SLOTS')
  File "/home/cbuisson/env-cache-benchmark/local/lib/python2.7/site-packages/rediscluster/utils.py", line 101, in inner
    return func(*args, **kwargs)
  File "/home/cbuisson/env-cache-benchmark/local/lib/python2.7/site-packages/rediscluster/client.py", line 272, in execute_command
    return self._execute_command_on_nodes(node, *args, **kwargs)
  File "/home/cbuisson/env-cache-benchmark/local/lib/python2.7/site-packages/rediscluster/client.py", line 357, in _execute_command_on_nodes
    res[node["name"]] = self.parse_response(connection, command, **kwargs)
  File "/home/cbuisson/env-cache-benchmark/local/lib/python2.7/site-packages/redis/client.py", line 587, in parse_response
    return self.response_callbacks[command_name](response, **options)
  File "/home/cbuisson/env-cache-benchmark/local/lib/python2.7/site-packages/rediscluster/utils.py", line 137, in parse_cluster_slots
    'master': fix_server(*master),
TypeError: fix_server() takes exactly 2 arguments (3 given)
```

**Comment by c-buisson** [comment]
I can *fixed* the issue by editing the `master` and `slaves` variables in the `utils.py` file:

```
    slots = {}
    for slot in resp:
        start, end, master = slot[:3]
        slaves = slot[3:]
        print "MASTER: {0}".format(master)
        master = ['10.0.0.1', 10000]
        print "MASTER edited: {0}".format(master)
        print "SLAVES: {0}".format(slaves)
        slaves = [['10.0.0.4`, 10000]]
        print "SLAVES edited: {0}".format(slaves)
        slots[start, end] = {
            'master': fix_server(*master),
            'slaves': [fix_server(*slave) for slave in slaves],
        }

    return slots
```

Now when I run `cluster_slots` I am getting this:

```
MASTER: ['10.0.0.2', 10000L, 'ffd36d8d7cb10d813f81f9662a835f6beea72677']
MASTER edited: ['10.0.0.1', 10000]
SLAVES: [['10.0.0.5', 10000L, '5c15b69186017ddc25ebfac81e74694fc0c1a160']]
SLAVES edited: [['10.0.0.4', 10000]]
MASTER: ['10.0.0.3', 10000L, '069cda388c7c41c62abe892d9e0a2d55fbf5ffd5']
MASTER edited: ['10.0.0.1', 10000]
SLAVES: [['10.0.0.6', 10000L, 'dc152a08b4cf1f2a0baf775fb86ad0938cb907dc']]
SLAVES edited: [['10.0.0.4', 10000]]
MASTER: ['10.0.0.1', 10000L, '3588b4cf9fc72d57bb262a024747797ead0cf7ea']
MASTER edited: ['10.0.0.1', 10000]
SLAVES: [['10.0.0.4', 10000L, 'a72c02c7d85f4ec3145ab2c411eefc0812aa96b0']]
SLAVES edited: [['10.0.0.4', 10000]]
// I am printing the content of r.client_slots() below
SLOTS: {(10923L, 16383L): {'master': ('10.0.0.1', 10000), 'slaves': [('10.0.0.4', 10000)]}, (5461L, 10922L): {'master': ('10.0.0.1', 10000), 'slaves': [('10.0.0.4', 10000)]}, (0L, 5460L): {'master': ('10.0.0.1', 10000), 'slaves': [('10.0.0.4', 10000)]}}
```
As you can see, the `master` and `slaves` variables are not only composed of the `host` and `port` but also of their own IDs.

**Comment by Grokzen** [comment]
@c-buisson I looked this over some and i can kinda confirm what you are seeing and that it is a breaking change that must have been introduced in some later version of redis 3.2.x.

Currently in my docker instance i get the following when i run the `CLUSTER SLOTS` command

```
127.0.0.1:7000> cluster slots
1) 1) (integer) 10923
   2) (integer) 16383
   3) 1) "172.17.0.2"
      2) (integer) 7002
   4) 1) "172.17.0.2"
      2) (integer) 7005
2) 1) (integer) 5461
   2) (integer) 10922
   3) 1) "172.17.0.2"
      2) (integer) 7001
   4) 1) "172.17.0.2"
      2) (integer) 7004
3) 1) (integer) 0
   2) (integer) 5460
   3) 1) "172.17.0.2"
      2) (integer) 7000
   4) 1) "172.17.0.2"
      2) (integer) 7003
```

and that kinda confirms the thing you said in your last post that it now adds the ID. It should be a rather easy fix to make it compatible again with the version you are running.

**Comment by Grokzen** [comment]
@c-buisson Closing this as fixed with this commit 9f9baeced12b7abac53b3680ea3f5a1b16a38038

See the tests that validates that it still works with the old and new response format.

**Comment by c-buisson** [comment]
Thanks for fixing this bug @Grokzen!
